packages/labs/ssr/src/lib/render.ts

```diff
@@ -43,5 +70,96 @@ export function* render(
   if (isTemplateResult(value)) {
     hydratable = isHydratable(value);
   }
-  yield* renderValue(value, renderInfo as RenderInfo, hydratable);
+  return renderValue(value, renderInfo as RenderInfo, hydratable);
+}
+
+type InternalRenderResultIterator = Iterator<string | Thunk>;
+
+/**
+ * Wraps a ThunkedRenderResult to implement a RenderResult.
+ */
+export class RenderResultIterator
+  implements Iterator<string | Promise<RenderResult>>
+{
+  /**
+   * A stack of open iterators.
+   *
+   * As the ThunkedRenderResult thunks are called, they may yield new arrays of
+   * strings and/or thunks. We push the iterators for these arrays onto the
+   * stack and consume them depth-first.
+   */
+  private _iterators: Array<InternalRenderResultIterator>;
+  private _waiting = false;
+
+  constructor(result: ThunkedRenderResult) {
+    this._iterators = [result[Symbol.iterator]()];
+  }
+
+  next(): IteratorResult<string | Promise<RenderResult>, unknown> {
+    if (this._waiting) {
+      throw new Error(
+        'Cannot call next() while waiting for a Promise to resolve'
+      );
+    }
+    const iterator = this._iterators.at(-1);
+    if (iterator === undefined) {
+      return {done: true, value: undefined};
+    }
+
+    // Get the next value from the current iterator
+    const result = iterator.next();
+    if (result.done) {
+      this._iterators.pop();
+      return this.next();
+    }
+    let value: string | Thunk | ReturnType<Thunk> = result.value;
+
+    // If the value is a string, return the result as-is:
+    if (typeof value === 'string') {
+      return result as IteratorResult<string, unknown>;
+    }
+
+    // Otherwise, it's a thunk. Trampoline to fully evaluate thunks:
+    while (typeof value === 'function') {
+      value = value();
+    }
+
+    // If the value is undefined, return the next value:
+    if (value === undefined) {
+      return this.next();
+    }
+
+    // If the value is a string, return a new iterator result:
+    if (typeof value === 'string') {
+      return {done: false, value};
+    }
+
+    // If the value is an array, push a new iterator for it onto the stack, and
+    // recurse to start consuming it:
+    if (Array.isArray(value)) {
+      this._iterators.push(value[Symbol.iterator]());
+      return this.next();
+    }
+
+    // The value is a Promise. Convert to a Promise<RenderResult>:
+    this._waiting = true;
+    return {
+      done: false,
+      value: value.then((r) => {
+        if (typeof r === 'string') {
+          return r;
```

Comment:
This might be a misunderstanding on my part, but should `this._waiting` not also be set to false in this early return?

Reply:
good catch!